satpy/dataset/dataid.py

```diff
@@ -732,21 +481,85 @@ def create_filtered_query(dataset_key, filter_query):
-    except AttributeError:
-        if isinstance(dataset_key, str):
-            ds_dict = {"name": dataset_key}
-        elif isinstance(dataset_key, numbers.Number):
-            ds_dict = {"wavelength": dataset_key}
-        else:
-            raise TypeError("Don't know how to interpret a dataset_key of type {}".format(type(dataset_key)))
+    elif isinstance(dataset_key, str):
+        ds_dict = {"name": dataset_key}
+    elif isinstance(dataset_key, numbers.Number):
+        ds_dict = {"wavelength": dataset_key}
+    else:
+        raise TypeError("Don't know how to interpret a dataset_key of type {}".format(type(dataset_key)))
     return ds_dict
+
+
+def update_id_with_query(orig_id: DataID, query: DataQuery) -> DataID:
+    """Update a DataID with additional info from a query used to find it."""
+    query_dict = query.to_dict()
+    if not query_dict:
+        return orig_id
+
+    new_id_dict = orig_id.to_dict()
+    orig_id_keys = orig_id.id_keys
+    for query_key, query_val in query_dict.items():
+        # XXX: What if the query_val is a list?
+        if new_id_dict.get(query_key) is None:
+            new_id_dict[query_key] = query_val
+    # don't replace ID key information if we don't have to
+    id_keys = orig_id_keys if all(key in orig_id_keys for key in new_id_dict) else default_id_keys_config
+    new_id = DataID(id_keys, **new_id_dict)
+    return new_id
+
+
+def _keys_to_compare(sdict: dict, odict: dict, o_is_id: bool, shared_keys: bool) -> set:
+    keys_to_match = set(sdict.keys())
+    if not o_is_id:
+        # if another DataQuery, then compare both sets of keys
+        keys_to_match |= set(odict.keys())
+    if shared_keys:
+        # only compare with the keys that both objects share
+        keys_to_match &= set(odict.keys())
+    return keys_to_match
+
+
+def _compare_key_equality(sdict: dict, odict: dict, key: str, o_is_id: bool) -> bool:
+    if key not in sdict:
+        return False
+    sval = sdict[key]
+    if sval == "*":
+        return True
+
+    if key not in odict:
+        return False
+    oval = odict[key]
+    if oval == "*":
+        # Gotcha: if a DataID contains a "*" this could cause
+        #    unexpected matches. A DataID is not expected to use "*"
+        return True
+
+    return _compare_values(sval, oval, o_is_id)
+
+
+def _compare_values(sval: Any, oval: Any, o_is_id: bool) -> bool:
+    if isinstance(sval, list) or isinstance(oval, list):
+        # multiple options to match
+        if not isinstance(sval, list):
+            # query to query comparison, make a list to iterate over
+            sval = [sval]
+        if o_is_id:
+            return oval in sval
+
+        # we're matching against a DataQuery who could have its own list
+        if not isinstance(oval, list):
+            oval = [oval]
+        s_in_o = any(_sval in oval for _sval in sval)
+        o_in_s = any(_oval in sval for _oval in oval)
+        return s_in_o or o_in_s
+    return oval == sval
```

Comment:
It’s probably my doing, but could we take the opportunity to find better/longer names for sval, oval, sdict, odict, etc?

Reply:
I really thought I answered a bunch of these comments already, but apparently not...

Yes, I can try to find better names. I think `s` is self or maybe source? And `o` is other. I'm not entirely upset with them as they are now, but yeah not super clear.